clang/tools/archgen-mlir/ArchGenMLIR.cpp

```diff
@@ -0,0 +1,927 @@
+      state.builder.setInsertionPointToStart(op->getBlock());
+      mlir::Value argVal =
+          state.builder
+              .create<LLVM::LoadOp>(op.getLoc(),
+                                    func.getBody().getArgument(param_idx))
+              .getRes();
+      op->replaceAllUsesWith(mlir::ValueRange{argVal});
+
+      /// Order matters.
+      maybeDelete.push_back(op);
+      maybeDelete.push_back(constantInt);
+    });
+    for (auto *op : maybeDelete)
+      if (op->use_empty())
+        op->erase();
+  }
+
+  mlir::LogicalResult runMLIROptimizationAndLowering() {
+    {
+      mlir::PassManager pm(&state.ctx);
+      pm.addPass(mlir::createInlinerPass());
+      if (mlir::failed(pm.run(state.module.get())))
+        return mlir::failure();
+    }
+    rewriteParameters();
+
+    {
+      mlir::PassManager pm(&state.ctx);
+      pm.addPass(mlir::createCanonicalizerPass());
+      pm.addPass(mlir::createCSEPass());
+      if (mlir::failed(pm.run(state.module.get())))
+        return mlir::failure();
+    }
+
+    /// This at this point the front-end and cleanup is done
+    if (mlir::failed(outputMLIR(MLIROutput, state.module.get())))
+      return mlir::failure();
+
+    if (StopAtApprox)
+      return mlir::success();
+
+    {
+      mlir::PassManager pm(&state.ctx);
+      // pm.enableIRPrinting();
+
+      /// The Core of ArchGen, will transform the approx dialect describing the
+      /// expression to be approximated into the logic to approximate the
+      /// expression
+      pm.addPass(approx::createLowerApproxPass());
+      /// When debugging it is a lot more useful to see the source locations form
+      /// here then from the eval call in the .cpp
+      pm.addPass(mlir::createLocationSnapshotPass());
+
+      /// These are not yet correct so they are disabled for now
+      pm.addPass(mlir::createCanonicalizerPass());
+      pm.addPass(mlir::createCSEPass());
+      pm.addPass(fixedpt::createConvertFixedPtToArithPass());
+      pm.addPass(mlir::createReconcileUnrealizedCastsPass());
+      pm.addPass(mlir::createCanonicalizerPass());
+      pm.addPass(mlir::createCSEPass());
+      pm.addPass(mlir::arith::createArithBufferizePass());
+      pm.addNestedPass<mlir::func::FuncOp>(mlir::createTensorBufferizePass());
+      pm.addPass(mlir::func::createFuncBufferizePass());
+      pm.addNestedPass<mlir::func::FuncOp>(mlir::bufferization::createFinalizingBufferizePass());
+      pm.addPass(mlir::createMemRefToLLVMConversionPass());
+      pm.addPass(mlir::createConvertIndexToLLVMPass());
+      pm.addPass(mlir::createArithToLLVMConversionPass());
+      pm.addPass(mlir::createConvertIndexToLLVMPass());
+      pm.addPass(mlir::createConvertFuncToLLVMPass());
+      pm.addPass(mlir::createReconcileUnrealizedCastsPass());
+      if (mlir::failed(pm.run(state.module.get())))
+        return mlir::failure();
+    }
+
+    return mlir::success();
+  }
+
+  mlir::LogicalResult mergeMLIRintoLLVMIR() {
```

Comment:
This function is crucial and you should explain what it does, in the ArchGenMLIR context of course.